spring-kafka/src/main/java/org/springframework/kafka/listener/ConcurrentMessageListenerContainer.java

```diff
@@ -380,30 +389,47 @@ protected void doStop(final Runnable callback, boolean normal) {
 
 	@Override
 	public void childStarted(MessageListenerContainer child) {
-		this.startedContainers.incrementAndGet();
+		this.lifecycleLock.lock();
+		try {
+			if (this.containers.contains(child) || this.stoppedContainers.contains(child)) {
+				this.startedContainers.incrementAndGet();
```

Comment:
This logic is somehow does not compile in my head.
Why do we need to track those `stoppedContainers` at all?
Does not look like that brings any benefit over `setFenced()` introduced before.

In my feeling the `this.containers` collection is totally enough to track all the lifecycle of children if their `start/stop` is done manually. 
Well, I still see a performance benefit via `this.startedContainers` counter, but this `stoppedContainers` smells like an overhead.

Reply:
When the container stop is called, all the `this.containers` are cleared. Assume not all containers really called childstopped.  Now, earlier child containers would invoke the childstopped API. How can we know that these child containers really belong to ConcurrentContainer for the active run or just stopped run. So, even though `stop` is called on ConcurrentContainer, we need to maintain the previous containers untill a second run starts or all the child containers are really stopped.

`this.stoppedContainers` would be useful here.

Reply:
Right. So, we must not do that `this.containers.clear();` in the end of `ConcurrentMessageListenerContainer.doStop(final Runnable callback, boolean normal)`.
And reset it when we do a new `ConcurrentMessageListenerContainer.start()`.
This way we would always have a track of children until we go fresh start 😄